README.md
# foodunits

Validate and convert food units with ease!

## Installation
You can install foodunits using pip:
```bash
$ pip install foodunits
```

## Usage
`foodunits` provides functionality to validate and convert food units. It supports both imperial and metric units, including common units such as "liter," "fluid ounce," "pint," "gram," "pound," as well as specific food-related units like "slice," "can," "bottle," and more.

### Validation
The validation feature checks whether a string represents a valid food unit. It is case, space, singular-plural, and full name-symbol insensitive. Only the spelling matters. It returns a boolean value indicating the validity of the unit. Here are some examples:

"Value + Unit" type strings can also be validated, such as "5 mls" or "five milliliters".

```python
from foodunits import units_validator

units_validator("fluid ounce")  # True
units_validator("fl oz")  # True
units_validator("5 fl ozs")  # True
units_validator("5 fl  ozs ")  # True
units_validator("five fl ozs")  # True
units_validator("flud ounce")  # False
units_validator("f oz")  # False
units_validator("5 l oz")  # False
units_validator("fiv fl ozs")  # False
```

### Conversation
The conversion feature allows you to convert between volumetric and mass food units, both in imperial and metric systems. For example:

```python
from foodunits import units_convertor

units_convertor("1 fluid ounce", to_unit="ml")
# Output: {"converted value": 29.573, "unit": "ml"}

units_convertor("1 1/2 fluid ounce", to_unit="ml")
# Output: {"converted value": 44.360, "unit": "ml"}
```

If you want to convert between volume and mass and have the correct conversion based on the food ingredient, you can provide the ingredient name. Currently, the package supports 100+ major ingredients. If the ingredient you input is not supported, you can provide a value for the "ingredient_density" argument. Here's an example:
```python
units_convertor("2.5", to_unit="g", from_unit="fl oz", ingredient="skimmed milk")
# Output: {"converted value": 76.152, "unit": "g"}
```

For cookery units like cup, tablespoon, and teaspoon, which depend on the country, you need to specify the country name or code:
"2.5 cups", "skimmed milk", "United States" -> "g"
```python
units_convertor("2.5 cups", to_unit="g", ingredient="skimmed milk", country="United States")
# Output: {"converted value": 618.0, "unit": "g"}
```

Note: The decimal parameter can be used to specify the number of decimal places in the converted value.

Make sure to import the relevant functions from the foodunits package to use them in your code.

## Contributing

Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.

## License

`foodunits` was created by Ben Zhang. It is licensed under the terms of the MIT license.

## Credits

`foodunits` was created with [`cookiecutter`](https://cookiecutter.readthedocs.io/en/latest/) and the `py-pkgs-cookiecutter` [template](https://github.com/py-pkgs/py-pkgs-cookiecutter). Inspired by [`validators`](https://github.com/python-validators/validators) and [`UnitConverter`](https://github.com/mattgd/UnitConverter)


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.0 (14/07/2023)

- First release of `foodunits`!


tests/test_convertor.py
"""Test food unit convertor"""
# -*- coding: utf-8 -*-
import pytest
from foodunits import units_convertor
from foodunits.exceptions import ConversionFailure

def test_convert_same_unit():
    result = units_convertor(1, "ml", "ml")
    assert result == {"converted value": 1, "unit": "ml"}


@pytest.mark.parametrize(
    "value, to_unit, from_unit, country, decimal_places, expected_result",
    [
        (1, "fl oz", "ml", "US", 3, {"converted value": 0.034, "unit": "fl oz"}), # Standard input
        ("1 ml", "fl oz", None, "US", 3, {"converted value": 0.034, "unit": "fl oz"}), # Value + unit
        ("1 ml.", " fl.  oz. ", None, "US", 3, {"converted value": 0.034, "unit": "fl oz"}), # Punctuations etc.
        ("1 fluid ounce. ", " ml. ", None, "US", 3, {"converted value": 29.573, "unit": "ml"}), # Full name unit.
        ("1. 5 fluid ounce. ", " ml.", None, "US", 3, {"converted value": 44.360, "unit": "ml"}), # Float string
        ("1 1/2 fluid ounce. ", "ml. ", None, "US", 3, {"converted value": 44.360, "unit": "ml"}), # Fraction string
    ],
)
def test_convert_volume_units(value, to_unit, from_unit, country, decimal_places, expected_result):
    result = units_convertor(value, to_unit, from_unit, country=country, decimal_places=decimal_places)
    assert result == expected_result

@pytest.mark.parametrize(
    "value, to_unit, from_unit, country, decimal_places, expected_result",
    [
        (2.5, "gram", "pounds", "US", 3, {"converted value": 1133.980, "unit": "g"}), # Standard input
        ("2.5 lbs", "gram", None, "US", 3, {"converted value": 1133.980, "unit": "g"}), # Value + unit
        ("2 1/2 ounces", "grams", None, "US", 3, {"converted value": 70.874, "unit": "g"}), # Punctuations etc.
        ("2.5 stones", "kg", None, "US", 3, {"converted value": 15.876, "unit": "kg"}), # Full name unit.
        ("2. 5 drams", "g", None, "US", 3, {"converted value": 4.430, "unit": "g"}), # Float string
        (" 2.5 grains. ", "grams ", None, "US", 3, {"converted value": 0.162, "unit": "g"}), # Fraction string
    ],
)
def test_convert_mass_units(value, to_unit, from_unit, country, decimal_places, expected_result):
    result = units_convertor(value, to_unit, from_unit, country=country, decimal_places=decimal_places)
    assert result == expected_result

@pytest.mark.parametrize(
    "value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result",
    [
        (2.5, "g", "fl oz", "water", None, "US", 3, {"converted value": 73.934, "unit": "g"}), # Standard input
        (73.934, "fl oz", "g", "water", None, "US", 3, {"converted value": 2.5, "unit": "fl oz"}),
        (73.934, "fl oz", "kg", "water", None, "US", 3, {"converted value": 2500.004, "unit": "fl oz"}),
        ("2.5 fl oz", "g", None, "skimme milk", None, "US", 3, {"converted value": 76.152, "unit": "g"}), # typo in ingredient
        ("2.5 pint", "g", None, "whole milk", 1.05, "US", 3, {"converted value": 1242.087, "unit": "g"}), # ingredient density provided.
    ],
)
def test_convert_mass_volume_units(value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result):
    result = units_convertor(
        value, to_unit, from_unit,
        ingredient=ingredient,
        ingredient_density=ingredient_density,
        country=country,
        decimal_places=decimal_places
    )
    assert result == expected_result

@pytest.mark.parametrize(
    "value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result",
    [
        (2.5, "ml", "cup", None, None, "US", 1, {"converted value": 600.0, "unit": "ml"}), # Standard input
        (2.5, "ml", "cup", None, None, "metric", 3, {"converted value": 625.0, "unit": "ml"}),
        ("2.5 cups", "fl oz", None, "water", None, "US", 3, {"converted value": 20.288, "unit": "fl oz"}),# cups to imperial
        ("2.5", "g", "cups", "skimmed milk", None, "united states", 3, {"converted value": 618.0, "unit": "g"}), # cups to mass
        ("2.5", "g", "cups", "skimmed milk", 1.1, "united states", 3, {"converted value": 660.0, "unit": "g"}), # cups to mass, with customized density
        ("600", "cup", "mls", "water", None, "united states", 3, {"converted value": 2.5, "unit": "cup"}), # metric to cups
        (20.288, "cup", "fl oz", "water", None, "united states", 3, {"converted value": 2.5, "unit": "cup"}), # imperial to cups
        (618, "cup", "g", "skimmed milk", None, "united states", 3, {"converted value": 2.5, "unit": "cup"}), # mass to cups
        ("2.5", "g", "tsps", "skimmed milk", None, "united states", 3, {"converted value": 12.692, "unit": "g"}), # teaspoon
        ("2.5", "g", "tbsps", "skimmed milk", None, "united states", 3, {"converted value": 38.076, "unit": "g"}), # tablespoon
    ],
)
def test_physical_container_units(value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result):
    result = units_convertor(
        value, to_unit, from_unit,
        ingredient=ingredient,
        ingredient_density=ingredient_density,
        country=country,
        decimal_places=decimal_places
    )
    assert result == expected_result

@pytest.mark.parametrize(
    "value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result",
    [
        ("missing value", "fl. oz", "ml", None, None, "US", 1,  ConversionFailure), # invalid value
        (2.5, "foo_to_unit", "foo_from_unit", None, None, "metric", 3,  ConversionFailure),
        (2.5, "g", "ml", "foo_ingredient", None, "US", 3,  ConversionFailure), # wrong ingredient
        (2.5, "g", "ml", None, None, "US", 3,  ConversionFailure), # missing ingredient
        (2.5, "tsp", "cup", None, None, "US", 3,  ConversionFailure), # can not convert between cup, tsp, tbsp.
        (2.5, "ml", "cup", "water", None, None, 3,  ConversionFailure), # missing country.
    ],
)
def test_convert_failure(value, to_unit, from_unit, ingredient, ingredient_density, country, decimal_places, expected_result):
    with pytest.raises(expected_result):
        units_convertor(
            value,
            to_unit,
            from_unit,
            ingredient=ingredient,
            ingredient_density=ingredient_density,
            country=country,
            decimal_places=decimal_places,
        )


tests/test_foodunits.py
from foodunits import foodunits


tests/test_validator.py
"""Test food unit validator"""
# -*- coding: utf-8 -*-
import pytest
from foodunits import units_validator
from foodunits.exceptions import ValidationFailure

@pytest.mark.parametrize(
    ("value",),
    [
        ("cup",),
        ("ounces",),
        ("Fluid Ounce",),
        ("slices",),
        ("2 mls",),
        ("2mls",),
        ("2 fl ozs",),
        ("one hundred and 2 fl ozs",),
    ],
)
def test_returns_true_on_valid_food_unit(value: str):
    """Test returns true on valid food unit."""
    assert units_validator(value)

@pytest.mark.parametrize(
    ("value",),
    [
        ("",),
        ("not_a_unit",),
        ("5mlls",),
    ],
)
def test_returns_failed_validation_on_invalid_food_unit(value: str):
    """Test returns failed validation on invalid food unit."""
    assert isinstance(units_validator(value), ValidationFailure)


src/foodunits/base.py
from foodunits.utils.units import Convert_Dict
from foodunits.exceptions import ConversionFailure

class FoodUnitConvertor:
    """Class contains mainly food conversion functions."""
    # Assign the available dictionary
    metric_dict = Convert_Dict.metric_dict()
    imperial_vol_dict = Convert_Dict.imperial_vol_dict()
    imperial_mass_dict = Convert_Dict.imperial_mass_dict()
    physical_container_unit = {
        "cup": Convert_Dict.cup_by_country_dict(),
        "teaspoon": Convert_Dict.teaspoon_by_country_dict(),
        "tablespoon": Convert_Dict.tablespoon_by_country_dict()
    }

    def __init__(
        self,
        value,
        units_from: str,
        units_to: str,
        density: float = None,
        country: str = "US",
        decimal_places: int = None,
        **kwargs
    ):
        """Initialization"""
        self.value = value
        self.units_from = units_from
        self.units_to = units_to
        self.decimal_places = decimal_places
        self.density = density
        self.country = country
        self.metric_units_from = units_from[:-1] if len(units_from) > 1 else None
        self.metric_units_to = units_to[:-1] if len(units_to) > 1 else None
        self.__dict__.update(kwargs)

    def check_metric_imperial(self):
        """
        Conversion between metric and imperial units.
        """
        converted_value = None
        # Metric and Imperial volumes （metric volumetric units should be l and m³)
        if self.metric_units_from in self.metric_dict and self.units_to in self.imperial_vol_dict:
            metric_base = self.metric_dict.get(self.metric_units_from, None)
            imperial_base = self.imperial_vol_dict.get(self.units_to, None)
            converted_value = self.metric_to_imperial(self.value, metric_base, imperial_base, 33.814)
            # Mass to volume
            if "g" in self.units_from:
                converted_value = self.volume_mass_conversion(converted_value, self.density, vol_to_mass=False)

        elif self.units_from in self.imperial_vol_dict and self.metric_units_to in self.metric_dict:
            imperial_base = self.imperial_vol_dict.get(self.units_from, None)
            metric_base = self.metric_dict.get(self.metric_units_to, None)
            converted_value = self.imperial_to_metric(self.value, metric_base, imperial_base, 0.0295735)
            # Volume to mass
            if "g" in self.units_to:
                converted_value = self.volume_mass_conversion(converted_value, self.density, vol_to_mass=True)

        # Metric and Imperial masses （metric mass units should be g, but exception is l or m³)
        elif self.metric_units_from in self.metric_dict and self.units_to in self.imperial_mass_dict:
            metric_base = self.metric_dict.get(self.metric_units_from, None)
            imperial_base = self.imperial_mass_dict.get(self.units_to, None)
            # Volume to mass
            if "l" in self.units_from:
                converted_value = self.metric_to_imperial(
                    self.volume_mass_conversion(self.value, self.density, vol_to_mass=True),
                    metric_base, imperial_base, 0.00220462
                )
            else:
                converted_value = self.metric_to_imperial(self.value, metric_base, imperial_base, 0.00220462)

        elif self.units_from in self.imperial_mass_dict and self.metric_units_to in self.metric_dict:
            imperial_base = self.imperial_mass_dict.get(self.units_from, None)
            metric_base = self.metric_dict.get(self.metric_units_to, None)
            converted_value = self.imperial_to_metric(self.value, metric_base, imperial_base, 453.592)
            # Mass to volume
            if "l" in self.units_to:
                converted_value = self.volume_mass_conversion(converted_value, self.density, vol_to_mass=False)

        if converted_value:
            return {
                "converted value": round(converted_value, self.decimal_places),
                "unit": self.units_to
            }
        else:
            return False

    def check_physical_container_unit(self):
        """
        Conversion from or to physical container units, e.g. cup to ml, or vice versa.
        """
        if not self.units_from in self.physical_container_unit and not self.units_to in self.physical_container_unit:
            return False
        #TODO Add converstion from cup to teaspoon/tablespoon etc.
        if self.units_from in self.physical_container_unit and self.units_to in self.physical_container_unit:
            raise ConversionFailure(
                    f"""
                    Can not convert between cup, teaspoon and tablespoon yet. To be added...
                    """
                )

        converted_value = None
        if self.units_from in self.physical_container_unit.keys():
            country_base = self.physical_container_unit[self.units_from].get(self.country, None)

            # No country is found
            if not country_base:
                raise ConversionFailure(
                    f"""
                    The converted units involve physical containers, such as cup, teaspoon, or tablespoon, which depend on the country.
                    Provide the accepted value from the followings(or corresponding full name) for the "country" argument:
                        {self.physical_container_unit[self.units_from].keys()}
                    """
                )

            if self.units_to in self.imperial_vol_dict:
                imperial_base = self.imperial_vol_dict.get(self.units_to, None)
                # convert cup etc. to liter first, then other imperial units
                converted_ml = self.value * country_base
                converted_value =  self.metric_to_imperial(converted_ml, 1, imperial_base, 33.814)

            elif self.metric_units_to in self.metric_dict:
                metric_base = self.metric_dict.get(self.metric_units_to, None)
                converted_value = country_base * self.value / metric_base

                if "g" in self.units_to:
                    converted_value = self.volume_mass_conversion(converted_value, self.density, vol_to_mass=True)

        if self.units_to in self.physical_container_unit.keys():
            country_base = self.physical_container_unit[self.units_to].get(self.country, None)

            # No country is found
            if not country_base:
                raise ConversionFailure(
                    f"""
                    The converted units involve physical containers, such as cup, teaspoon or tablespoon which is country-matters.
                    Provide the accepted value for argument "country":
                        {self.physical_container_unit[self.units_to].keys()}
                    """
                )

            if self.units_from in self.imperial_vol_dict:
                # convert imperial units to liter first, then cup etc.
                imperial_base = self.imperial_vol_dict.get(self.units_from, None)
                converted_value = self.imperial_to_metric(self.value, 1, imperial_base, 0.0295735) / country_base

            elif self.metric_units_from in self.metric_dict:
                metric_base = self.metric_dict.get(self.metric_units_from, None)
                # convert to volume first
                if "g" in self.units_from:
                    converted_value = self.volume_mass_conversion(self.value, self.density, vol_to_mass=False) * metric_base / country_base
                else:
                    converted_value = self.value * metric_base / country_base



        if converted_value:
            return {
                "converted value": round(converted_value, self.decimal_places),
                "unit": self.units_to
            }
        else:
            return False

    @staticmethod
    def volume_mass_conversion(value, density, vol_to_mass: bool = True):
        """
        Convert between mass and volume using a certain density.
        """
        if vol_to_mass:
            result = 1000 * value * density
        else:
            result = value / (1000 * density)
        return result

    @staticmethod
    def metric_to_imperial(value, metric_base, imperial_base, imperial_multiplier):
        """
        Convert from metric to imperial units.
        """
        return value * metric_base * imperial_multiplier / imperial_base

    @staticmethod
    def imperial_to_metric(value, metric_base, imperial_base, metric_multiplier):
        """
        Convert from imperial to metric units.
        """
        return value * imperial_base * metric_multiplier / metric_base

    @staticmethod
    def metric_to_metric(value, metric_base, metric_multiplier):
        """
        Convert between metric units.
        """
        return value * metric_base * metric_multiplier


src/foodunits/convertor.py
"""Module run food unit conversion"""
import re
import logging
from typing import Tuple, Dict, Any
from pattern.text.en import singularize
from foodunits.utils.utils import split_quantity_unit, validate_numeric_string, preprocess, get_ingredient_density, find_country
from foodunits.utils.units import UNITS, Convert_Dict
from foodunits.base import FoodUnitConvertor
from foodunits.exceptions import ConversionFailure


def can_convert(
    from_unit: str,
    to_unit: str,
    ingredient: str,
    ingredient_density: Tuple[int, float]
) -> Any:
    """
    Check if the given units can be converted. Currently only checks weight and volume units.
    Args:
        from_unit: The source unit to convert from
        to_unit: The target unit to convert to
        ingredient: If converting between mass and volume, ingredient or its density should be provided
        ingredient_density: If converting between mass and volume, ingredient or its density should be provided
    Returns:
        si_from_unit: The SI form of from_unit
        si_to_unit): The SI form of to_unit
    """
    acceptable_cats = ["weight", "volume"]
    _threshold =85

    # Check if units belong to acceptable categories
    from_unit_category = _find_unit_category(from_unit)
    to_unit_category = _find_unit_category(to_unit)
    if not from_unit_category or not to_unit_category or \
        from_unit_category["name"] not in acceptable_cats or \
        to_unit_category["name"] not in acceptable_cats:
        raise ConversionFailure(f"Both units should be in {acceptable_cats} category")

    if from_unit_category["name"] != to_unit_category["name"]:
        if not ingredient_density:
            ingredient_density = get_ingredient_density(
                ingredient,
                ingredient_dict = Convert_Dict.ml_to_g_by_ingredient_dict(),
                threshold = _threshold
            )
            if not ingredient_density:
                raise ConversionFailure(
                    """
                    Converstion between volume and mass.
                    Please checked the presence of arguement "ingredient" and any typos in it,
                    or specify the arguement "ingredient_density" to proceed converstion.
                    """
                )

    return get_si(from_unit), get_si(to_unit), ingredient_density


def get_si(unit: str) -> str:
    """
    Load the International System of Units (SI) string for the given unit.
    Args:
        unit: The unit for which the SI string should be loaded
    Returns:
        str: SI string or None if none was found
    """
    cat_unit = _find_unit(unit)
    if cat_unit:
        # return full name only for "cup", "teaspoon", "tablespoon"
        if cat_unit["name"] in ["cup", "teaspoon", "tablespoon"]:
            return cat_unit["name"]
        si = cat_unit["si"]
    return si if si else unit


def _find_unit_category(unit: str) -> Dict:
    """
    Internal: Load the category of one unit from the UNITS array.
    Args:
        unit: The unit which should be included in the '_internal_accepted_names'
            in a category of the UNITS array
    Returns:
        Dict: The category from the UNITS array or None if none was found
    """
    for category in UNITS:
        for cat_unit in category["units"]:
            if unit == cat_unit["name"] or unit == cat_unit["si"]:
                return category
    return None


def _find_unit(unit: str) -> Dict:
    """
    Internal: Load one specific unit from the UNITS array.
    Args:
        unit: The name of the unit (should be included in the '_internal_accepted_names' or 'name')
    Returns:
        Dict: The unit from the UNITS array or None if none was found
    """
    for category in UNITS:
        for cat_unit in category["units"]:
            if unit == cat_unit["name"] or unit == cat_unit["si"]:
                return cat_unit
    return None


def units_convertor(
    value: Tuple[str, int, float],
    to_unit: str,
    from_unit: str = None,
    ingredient: str = None,
    ingredient_density: float = None,
    country: str = None,
    decimal_places: int = None,
) -> Dict:
    """
    Convert the given value from the source unit to the target unit.
    Args:
        value: Value to convert, accepted forms include value only, or value + unit
               (e.g., "1 mls", just "1", 1, "one")
        from_unit: Source unit to convert from
        to_unit: Target unit to convert to
        ingredient: If converting between mass and volume, ingredient or its density should be present
        ingredient_density: If converting between mass and volume, ingredient or its density should be present
        country: Country for unit conversions (default: None)
        decimal_places: Number of decimal places for the converted value (default: None)
    Returns:
        Dict: Dictionary of converted value and unit
    """
    # Convert string input to value or value + unit
    if isinstance(value, str):
        cleaned_value = preprocess(value)
        valid, valid_value = validate_numeric_string(cleaned_value)
        if valid:
            value = valid_value
        else:
            value, unit_split = split_quantity_unit(cleaned_value)
            valid, valid_value = validate_numeric_string(value)
            from_unit = from_unit if from_unit else unit_split
            value = valid_value if valid else None
    # Check the converted value
    if not isinstance(value, (int, float)):
        raise ConversionFailure(
            f"""
            The input value {value} should be either int or float or convertable strings, such as:
                "5",
                "5.5",
                "5 fl ozs",
                "5 fluid ounces",
                "five fluid ounces",
                "5mls" etc.
            """
        )

    # Process unit; keep only alphabets and one space
    from_unit = singularize(re.sub(r"\s+", " ", re.sub(r'[^A-Za-z\s\d]+', '', from_unit)).strip()).lower()
    to_unit = singularize(re.sub(r"\s+", " ", re.sub(r'[^A-Za-z\s\d]+', '', to_unit)).strip()).lower()
    # Check if units can be converted
    try:
        from_unit, to_unit, ingredient_density = can_convert(from_unit, to_unit, ingredient, ingredient_density)
    except ConversionFailure as e:
        # Handle the specific custom error (ConversionFailure)
        logging.error("Conversion error: %s", str(e))
        raise
    except Exception as e:
        # Handle all other exceptions
        logging.error("Exception occurred: %s", str(e))
        raise

    # Extract the numeric value from the string
    if not decimal_places:
        if "." in str(value):
            decimal_places = len(str(value)[str(value).index('.') + 1:])
        else:
            decimal_places = 0

    # Return the value if units are the same
    if from_unit == to_unit:
        return {"converted value": value, "unit": get_si(to_unit)}

    # Get the country code
    country = find_country(country)

    convertor = FoodUnitConvertor(
        value, from_unit, to_unit,
        density=ingredient_density,
        country=country,
        decimal_places=decimal_places
    )
    responses = [
        convertor.check_metric_imperial(),  # Metric and Imperial units
        convertor.check_physical_container_unit()  # Cup and Teaspoon
    ]

    for response in responses:
        if response:
            return response

    # Return a default response if no conversions found
    return {"converted value": None, "unit": None}


src/foodunits/exceptions.py
"""Exceptions"""
from typing import Callable, Dict, Any

class ValidationFailure(Exception):
    """Exception class for validation failures."""

    def __init__(self, function: Callable[..., Any], arg_dict: Dict[str, Any], message: str = ""):
        """Initialize ValidationFailure."""
        if message:
            self.reason = message
        self.func = function
        self.__dict__.update(arg_dict)

    def __repr__(self):
        """Return the string representation of ValidationFailure."""
        return (
            f"ValidationFailure(func={self.func.__name__}, "
            + f"args={({k: v for (k, v) in self.__dict__.items() if k != 'func'})})"
        )

    def __str__(self):
        """Return the string representation of ValidationFailure."""
        return repr(self)

    def __bool__(self):
        """Return False for ValidationFailure."""
        return False

class ConversionFailure(Exception):
    """
    Exception class indicating that a conversion is not possible.
    """

    def __init__(self, reason: str):
        self.reason = reason

    def __repr__(self):
        """Return the string representation of ConversionFailure."""
        return f"ConversionFailure: {self.reason}"


src/foodunits/utils/units.py
"""Food unit and conversion rate dictionary"""
# -*- coding: utf-8 -*-
class Dictionary():
    """
    Convertion rate
    """
    # Metric and Imperial systems
    __metric_dict = {
        'E': 1000000000000000000,
        'P': 1000000000000000,
        'T': 1000000000000,
        'G': 1000000000,
        'M': 1000000,
        'k': 1000,
        'h': 100,
        'da': 10,
        None: 1,
        'd': .1,
        'c': .01,
        'm': .001,
        'μ': .000001,
        'n': .000000001,
        'p': .000000000001,
        'f': .000000000000001,
        'a': .000000000000000001
    }
    __imperial_vol_dict = {
        'gal': 128,
        'qt': 32,
        'pt': 16,
        'gi': 4,
        'fl oz': 1
    }
    __imperial_mass_dict = {
        't': 2240,
        'cwt': 112,
        'qr': 28,
        'qtr': 28,
        'st': 14,
        'lb': 1,
        'oz': .0625,
        'dr': .00390625,
        'gr': .00014285714
    }
    # cup to liter
    __cup_by_country_dict = {
        'metric': 0.25,
        'international': 0.25,
        'other': 0.25,
        'us': 0.24,
        'us legal': 0.24,
        'us customary': 0.2365882365,
        'ca': 0.25,
        'uk': 0.2841,
        'jp': 0.2,
        'kr': 0.2,
        'ru': 0.246,
    } # to be expend

    # teaspoon to liter
    __teaspoon_by_country_dict = {
        'metric': 0.005,
        'international': 0.005,
        'other': 0.005,
        'us': 0.00492892,
        'imperial': 0.005919,
    } # to be expend

    # tablespoon to liter
    __tablespoon_by_country_dict = {
        'metric': 0.015,
        'international': 0.0147868,
        'au': 0.02,
        'ca': 0.015,
        'uk': 0.015,
        'us': 0.0147868,
        'imperial': 0.0177582,
    } # to be expend

    # add custom conversion rate
    __ml_to_g_by_ingredient_dict = {
        "water": 1, # water
        "flour": 0.529, # flour
        "all-purpose flour": 0.529,
        "almond flour": 0.406,
        "bread flour": 0.55,
        "cake flour": 0.482,
        "coconut flour": 0.51,
        "corn flour": 0.65,
        "chickpea flour": 0.58,
        "oat flour": 0.64,
        "pastry flour": 0.448,
        "whole wheat flour": 0.478,
        "baking powder": 0.9,
        "baking soda": 0.69,
        "cream of tartar": 0.68,
        "sourdough starter": 0.972,
        "granulated sugar": 0.85, # sugars and sweetners
        "brown sugar": 0.95,
        "powdered sugar": 0.65,
        "honey": 1.42,
        "maple syrup": 1.35,
        "agave nectar": 1.4,
        "molasses": 1.45,
        "corn syrup": 1.38,
        "coconut sugar": 0.9,
        "stevia (powder)": 0.12,
        "butter": 0.911, # fats and oils
        "vegetable oil": 0.920,
        "olive oil": 0.920,
        "canola oil": 0.922,
        "coconut oil": 0.924,
        "sunflower oil": 0.924,
        "sesame oil": 0.929,
        "peanut oil": 0.924,
        "lard": 0.859,
        "shortening": 0.918,
        "almond milk": 1.04, # milk and cream
        "buttermilk": 1.01,
        "cashew milk": 1.03,
        "coconut milk": 1.02,
        "condensed milk": 1.28,
        "evaporated milk": 1.14,
        "half-and-half": 1.011,
        "heavy cream": 1.007,
        "light cream": 1.011,
        "oat milk": 1.03,
        "rice milk": 1.03,
        "semi-skimmed milk": 1.03,
        "skimmed milk": 1.03,
        "soy milk": 1.02,
        "whipping cream": 1.011,
        "whole milk": 1.03,
        "milk": 1.03,
        "almonds": 0.595,  # nuts and seeds
        "cashews": 0.585,
        "chia seeds": 0.555,
        "flaxseeds": 0.525,
        "hazelnuts": 0.655,
        "pecans": 0.625,
        "pine nuts": 0.595,
        "pistachios": 0.605,
        "poppy seeds": 0.605,
        "sesame seeds": 0.585,
        "sunflower seeds": 0.580,
        "walnuts": 0.595,
        "apples": 0.960, # fruits and vegs
        "bananas": 0.950,
        "blueberries": 0.840,
        "carrots": 0.810,
        "grapes": 0.810,
        "lemons": 0.960,
        "onions": 0.870,
        "oranges": 0.930,
        "potatoes": 0.775,
        "spinach": 0.760,
        "strawberries": 0.780,
        "tomatoes": 0.940,
        "watermelons": 0.960,
        "salt": 1.20, # other major ingredients
        "black pepper": 0.62,
        "garlic (minced)": 1.20,
        "onion (diced)": 0.80,
        "white vinegar": 1.04,
        "soy sauce": 1.03,
        "mustard": 1.12,
        "ketchup": 1.40,
        "mayonnaise": 0.94,
        "egg (whole)": 1.06,
        "bacon (cooked)": 0.70,
        "shrimp (peeled and deveined)": 0.90,
        "chicken breast": 0.99,
        "beef (ground)": 0.90,
        "pasta (uncooked)": 0.68,
        "rice (uncooked)": 0.68,
        "chocolate chips": 0.55,
        "vanilla extract": 1.03,
        "cocoa powder": 0.86
    } # to be expend

    def metric_dict(self):
        return self.__metric_dict

    def imperial_vol_dict(self):
        return self.__imperial_vol_dict

    def imperial_mass_dict(self):
        return self.__imperial_mass_dict

    def cup_by_country_dict(self):
        return self.__cup_by_country_dict

    def teaspoon_by_country_dict(self):
        return self.__teaspoon_by_country_dict

    def tablespoon_by_country_dict(self):
        return self.__tablespoon_by_country_dict

    def ml_to_g_by_ingredient_dict(self):
        return self.__ml_to_g_by_ingredient_dict



# UNITS list, all the values should be lower case
UNITS = [
    {
        "name": "weight",
        "units": [
            {"name": "gram", "si": "g"}, #metric
            {"name": "kilogram", "si": "kg"},
            {"name": "milligram", "si": "mg"},
            {"name": "ounce", "si": "oz"}, #imperial
            {"name": "pound", "si": "lb"},
            {"name": "ton", "si": "t"},
            {"name": "hundredweight", "si": "cwt"},
            {"name": "quarter", "si": "qr"},
            {"name": "stone", "si": "st"},
            {"name": "dram", "si": "dr"},
            {"name": "grain", "si": "gr"}
        ]
    },
    {
        "name": "volume",
        "units": [
            {"name": "liter", "si": "l"}, #metric
            {"name": "milliliter", "si": "ml"},
            {"name": "gallon", "si": "gal"}, #imperial
            {"name": "quart", "si": "qt"},
            {"name": "pint", "si": "pt"},
            {"name": "gill", "si": "gi"},
            {"name": "fluid ounce", "si": "fl oz"},
            {"name": "fluid oz", "si": "fl oz"},
            {"name": "fl ounce", "si": "fl oz"},
            {"name": "teaspoon", "si": "tsp"}, #other
            {"name": "tablespoon", "si": "tbsp"},
            {"name": "cup", "si": None}
        ]
    },
    {
        "name": "distance",
        "units": [
            {"name": "millimeter", "si": "mm"},
            {"name": "centimeter", "si": "cm"},
            {"name": "inch", "si": "in"}
        ]
    },
    {
        "name": "other",
        "units": [
            {"name": "slice", "si": None},
            {"name": "whole", "si": None},
            {"name": "dozen", "si": None},
            {"name": "pinch", "si": None},
            {"name": "dash", "si": None},
            {"name": "drop", "si": None},
            {"name": "sprinkle", "si": None},
            {"name": "handful", "si": None},
            {"name": "bunch", "si": None},
            {"name": "head", "si": None},
            {"name": "clove", "si": None},
            {"name": "fillet", "si": None},
            {"name": "portion", "si": None},
            {"name": "serving", "si": None},
            {"name": "piece", "si": None},
            {"name": "packet", "si": None},
            {"name": "can", "si": None},
            {"name": "jar", "si": None},
            {"name": "bottle", "si": None}
        ]
    }
]

Convert_Dict=Dictionary()
